### PR TITLE
JDK-8276892: Provide a way to emulate exceptional situations in FileManager when using JavadocTester

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a moderately simple addition to the `JavadocTester` world, to be able to modify the behavior of a file manager to throw exceptions as needed.

The bulk of the new functionality is in a new class, `TestJavaFileManagerBuilder`, which uses the builder pattern to allow a client to configure and build a file manager that can provide file objects for which the behavior can be overridden for one or more methods.  The expected use case is to throw an exception instead of calling the underlying delegate method.

`JavadocTester` is modified to allow a file manager to be provided when invoking `javadoc`. This requires some minor changes to the outermost javadoc tool classes, `Main` and `Start`.  Rather than add more `static` methods to `Main`, instance methods are now provided to make it easier to configure the values that will be passed to `Start`.    In `Start`, it was previously assumed that either the default file manager was being used or that _all_ paths would be configured directly in the file manager. The latter part of that assumption is reduced so that path-setting options (e.g. `--source-path`, `--class-path` etc) can be passed to the provided file manager. (Previously, they were silently ignored.) It is an error to pass path-like options as javadoc options to a file manager that does not support them.  However, since none of the changes are visible to any public API, this should not be an issue.

A new test is provided, with a few simple test cases. One is for direct use of the new file manager mechanism (without using `javadoc`). The other two illustrate how the feature can be used inside a `JavadocTester` call of `javadoc`.  Given that the feature is a relatively simple combination of predicates and proxies, it's not clear that we need a significant body of test cases.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276892](https://bugs.openjdk.java.net/browse/JDK-8276892): Provide a way to emulate exceptional situations in FileManager when using JavadocTester

### Reviewers
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to 9130979136da17fe1c071bffd119d6f9c1469a83

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6404/head:pull/6404` \
`$ git checkout pull/6404`

Update a local copy of the PR: \
`$ git checkout pull/6404` \
`$ git pull https://git.openjdk.java.net/jdk pull/6404/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6404`

View PR using the GUI difftool: \
`$ git pr show -t 6404`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6404.diff">https://git.openjdk.java.net/jdk/pull/6404.diff</a>

</details>

### CODE REVIEW

**Review by pavelrappo — CHANGES_REQUESTED** [review]
Initial comments.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/doclets.properties [hunk 1]
```diff
@@ -45,7 +45,7 @@ doclet.exception.write.file=Error writing file: {0}\n\
 \t({1})
 doclet.exception.read.resource=Error reading system resource: {0}\n\
 \t({1})
-doclet.internal.exception=An internal exception has occurred. \n\
+doclet.internal.exception=An internal exception has occurred.\n\
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/doclets.properties:48]: The message shows up in the new test.   Both the trailing space and the tabs caused problems in the string matching. I fixed the space here, as a one-off.   The tabs in this and related messages seem questionable and should maybe also be addressed, but that seemed too much out of scope for this work.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Main.java [hunk 2]
```diff
@@ -88,6 +88,65 @@ public static int execute(String[] args, PrintWriter outWriter, PrintWriter errW
...
+     */
+    public Main setFileManager(StandardJavaFileManager fileManager) {
+        this.fileManager = fileManager;
+        return this;
+    }
+
+    /**
+     * Runs javadoc with preconfigured values and a given set of arguments.
+     * Any errors will be reported to the error stream, or to {@link System#err}
+     * if not error stream has been specified with {@code setStreams}.
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Main.java:134]: Noted.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java [hunk 3]
```diff
@@ -371,9 +373,22 @@ public void javadoc(String... args) {
         StreamOutput sysErr = new StreamOutput(System.err, System::setErr);
 
         try {
+            /*
             exitCode = useStandardStreams
                     ? jdk.javadoc.internal.tool.Main.execute(args)              // use sysOut, sysErr
                     : jdk.javadoc.internal.tool.Main.execute(args, outOut.pw);  // default
+             */
+            jdk.javadoc.internal.tool.Main main = new jdk.javadoc.internal.tool.Main();
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/JavadocTester.java:381]: Thanks; no;  I'll remove it.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+ * However, it does mean that methods on file objects called internally by a
+ * file manager will not use the replacement functions.
+ *
+ * For now, the functions are simply grouped by predicate and then by method,
+ * and the group of methods used for a "test file object" is determined by the
+ * first predicate that matches.
+ * An alternative, more expensive, implementation would be to group the functions
+ * by method and predicate and then dynamically build the set of methods to be used for
+ * a file object by filtering the methods by their applicable predicate.
+ *
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java:66]: Noted.

**Comment by jonathan-gibbons** [comment]
Here is additional note/suggestion ...

The basic API for the builder is very, well, basic, using `Predicate<JFO>` and `Method` objects.
In time, we might want to add some convenience fields or methods to the builder to reduce call-site clutter from custom lambdas.

Early drafts of the code provided overloads for the `handle` method, but writing the tests showed it was hard to anticipate what might be required ... so providing constants or methods to use with the basic `handle` method may be a good way to go.

**Comment by pavelrappo** [comment]
After having looked at this change, I think I might have fallen prey to the famous "XY problem". Instead of asking on how to emulate error conditions in a black-box fashion first, I asked to provide white-box support for those through JavadocTester.

Just to make sure we're on the same page and aren't trying to solve a more complex problem than is required. Is there an easy way to make these methods throw: `javax.tools.JavaFileManager#getFileForInput` and `javax.tools.FileObject#getCharContent`?

**Comment by jonathan-gibbons** [comment]
> After having looked at this change, I think I might have fallen prey to the famous "XY problem". Instead of asking on how to emulate error conditions in a black-box fashion first, I asked to provide white-box support for those through JavadocTester.

I didn't see the question this way, and I thought the JBS issue was well stated.

Setting aside `JavadocTester` for a moment, there were two ways to solve the general problem of triggering exceptions to emulate weird conditions.

1. Build back doors into the main system `JavaFileManager` ... i..e. `JavacFileManager` (note `Javac`), such that the code can be triggered by backdoor access, either reflective, or breaking abstraction or system properties.
2. Provide a new file manager to do the job.

I prefer not to do the first ... I prefer not to build those backdoors into the main file manager that is routinely used everywhere. So that means to do the second.

Providing a new file manager to do the job means either using `javax.tools.DocumentationTool` directly, or else updating `JavadocTester`. The latter seems more desirable because of all the extra code in `JavadocTester.javadoc(...)` surrounding the actual tool invocation.    And it comes at the cost of tweaking the tool invocation.  It would be possible to have `JavadocTester` invoke `Start` directly without modification, but that would have required modifying _all_ existing tests, because of the need to use `com.sun.tools.javac.util.Context`.  Stepping away from that means modifying either `Main` or `Start` to accept a file manager directly, and of the two, it seemed more future-proof to adapt `Main` into a builder-style API, without otherwise changing any existing code paths.

The one surprise/disappointment was the "sort-of technical debt" I discovered in `Start` ... that file-manager options in the "command-line" args were ignored if a non-standard file manager was used. That seemed like a worthwhile issue to address here.

In terms of providing a custom file manager, the question is, "which file objects" and "which methods". I went through a series of iterations before deciding on using `Predicate<JavaFileObject>` to filter the file objects. For a while, the API used regular expressions applied to the name of the file object, but that runs into issues with platform-specific file separators. Yeah, we could use the URI, but equally, it's nicer to have access to the file object itself, for filtering directly on properties of the file object, including the possibility of `.equals()`.  As for "which methods", yeah, you suggested the two methods used to _read_ objects, but what if we want to write tests where the exceptions are thrown when _writing_ the object.

That all being said, I accept that the API as presented at this point may seem to be a sledgehammer, and it may be convenient to provide a coat of paint.   I like the general underlying architecture of the solution, but maybe we can come up with a method on the builder along the following lines:

    static StandardJavaFileManager throwExceptionWhenReading(StandardJavaFileManager fm, JavaFileObject jfo)

We can bike-shed on the name, and maybe there's a 3rd argument `Function<JavaFileObject, Throwable>` to create the exception to supply the exception that will be thrown.    And, there's clearly a family of possible overloads, such as replacing the `JavaFileObject` parameter with a `Predicate<JavaFileObject>` to make it easier to specify the behavior for more than one file object, if that should be useful, or `Predicate<String>` to specify a predicate on the file name, or last component of the filename.

To summarize, I don't think the change is as scary as it might appear at first, although it may be somewhat lacking in ease-of-use for common cases.  While the code of the builder may seem complicated, with its use of reflection and proxies, the entire class, the file is only 240 lines, including the legal header block and documentation comments.

> Just to make sure we're on the same page and aren't trying to solve a more complex problem than is required. Is there an easy way to make these methods throw: javax.tools.JavaFileManager#getFileForInput and javax.tools.FileObject#getCharContent?

TL;DR: Yes.   But I didn't know about the getFileForInput use case. I'll look to make that possible.

**Comment by jonathan-gibbons** [comment]
> TL;DR: Yes. But I didn't know about the getFileForInput use case. I'll look to make that possible.

There is an interesting difference between trapping  `javax.tools.JavaFileManager#getFileForInput` and `javax.tools.FileObject#getCharContent`.

For `getCharContent` we are filtering on the receiver of the method call.  For `getFileForInput` we would want to trap on the

**Comment by jonathan-gibbons** [comment]
(Sorry for the accidental closed/re-open)

**Comment by pavelrappo** [comment]
> For getFileForInput we would want to trap on the
That sentence ends abruptly.

**Comment by jonathan-gibbons** [comment]
> > For getFileForInput we would want to trap on the
> 
> That sentence ends abruptly.

I had meant to delete the comment, because I wasn't ready to finish it.; that must be how/why I accidentally closed the PR.   For now, I have completed the sentence, but generally, you can ignore the comment.  I'll follow up when I have more concrete ideas.

**Comment by pavelrappo** [comment]
I should've asked an even more general question before creating JDK-8276892: How can I *trigger* an error from particular operations on a FileManager and a FileObject returned from that FileManager?

A simulation through a custom implementation of these types would be a possible answer. However, providing a custom implementation is the white-box approach, and as far as I understand, JDK contributors traditionally prefer the black-box approach.

In the black-box approach, one seeks ways to set up an environment so that a desired effect is achieved indirectly. For example, to trigger IOException on an attempt to delete a file, I would ensure that the path does not exist or that the JVM has insufficient file-system permissions to access that path. Here's another example. To achieve socket blocking on write, I would fill it in with as much data as possible (SO_SNDBUF) while simultaneously ensuring that no one reads from it. 

If you say that a black-box approach is impossible or impractical here, I'll be fine with the white-box approach along the lines of what this PR suggests.

**Comment by pavelrappo** [comment]
> > After having looked at this change, I think I might have fallen prey to the famous "XY problem". Instead of asking on how to emulate error conditions in a black-box fashion first, I asked to provide white-box support for those through JavadocTester.
> 
> I didn't see the question this way, and I thought the JBS issue was well stated.
> 
> Setting aside `JavadocTester` for a moment, there were two ways to solve the general problem of triggering exceptions to emulate weird conditions.
> 
> 1. Build back doors into the main system `JavaFileManager` ... i..e. `JavacFileManager` (note `Javac`), such that the code can be triggered by backdoor access, either reflective, or breaking abstraction or system properties.
> 2. Provide a new file manager to do the job.
>
> [...]

I should've asked an even more general question before creating JDK-8276892: How can I _trigger_ an error from specific operations on a FileManager and a FileObject returned from that FileManager?

A simulation through a custom implementation of these types that this PR proposes would be one answer. However, such a simulation is the white-box approach, and as far as I understand, JDK contributors traditionally prefer the black-box approach.

In the black-box approach, one seeks ways to emulate the environment that would yield the desired effect. For example, to trigger IOException on an attempt to delete a file, I would ensure that the path does not exist or that the JVM has insufficient file-system permissions to access that path. Here's another example. To achieve socket blocking on write, I would fill it in with as much data as possible (SO_SNDBUF) while simultaneously ensuring that no one reads from it.

If you say that a black-box approach is impossible or impractical here, I'll be fine with the white-box approach along the lines of what this PR suggests.

**Comment by jonathan-gibbons** [comment]
> > > After having looked at this change, I think I might have fallen prey to the famous "XY problem". Instead of asking on how to emulate error conditions in a black-box fashion first, I asked to provide white-box support for those through JavadocTester.
> > 
> > 
> > I didn't see the question this way, and I thought the JBS issue was well stated.
> > Setting aside `JavadocTester` for a moment, there were two ways to solve the general problem of triggering exceptions to emulate weird conditions.
> > 
> > 1. Build back doors into the main system `JavaFileManager` ... i..e. `JavacFileManager` (note `Javac`), such that the code can be triggered by backdoor access, either reflective, or breaking abstraction or system properties.
> > 2. Provide a new file manager to do the job.
> > 
> > [...]
> 
> I should've asked an even more general question before creating JDK-8276892: How can I _trigger_ an error from specific operations on a FileManager and a FileObject returned from that FileManager?
> 
[...]
> 
> If you say that a black-box approach is impossible or impractical here, I'll be fine with the white-box approach along the lines of what this PR suggests.

I don't know of a way to reliably trigger the exceptions with the current primary file manager ... i.e. `JavacFileManager` (note the `c`).  You might be able to force exceptions by playing with file system permissions, but that may be problematic to do in a platform-independent way.

Here's a different way of looking at it, as an instance of a more general problem.  `JavaFileManager` is effectively a service-provider interface, to provide an interface between "files" and `javac`. At the time it was written, the only access to files was via `java.io.File` and there was interest in providing access to "virtual files" such as in-memory buffers, as may be found in an IDE.  (`JavaFileManager` appeared in JDK 6, and so predates the NIO `Path` and `FileSystem` APIs, which appeared later, in JDK 7;  if NIO had been available in JDK 6, we might not have written `JavaFileManager`).  So, as a service-provider interface, methods were declared to throw exceptions that might be expected to occur on as yet unknown implementations.  Thus, there is no guarantee that any specific implementation will need or be able to throw any of the declared exceptions.

For the situation that motivated this issue, the desire is to test/cover the `catch` blocks, so it does not seem so important to me how the exceptions are generated ... whether by tricking the raw `JavacFileManager` into throwing exceptions, or by wrapping it and using reflective proxy code, as here.   The one thing I take away is to maybe try and improve ease-of-use, perhaps with more convenience or utility methods.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+     * @throws IllegalArgumentException if the method is not declared in {@code JavaFileManager}
+     * @throws IllegalStateException if a handler is already registered for this method
+     */
+    public TestJavaFileManagerBuilder handle(Method method,
+                                             BiFunction<JavaFileManager, Object[], Throwable> handler) {
+        if (!JavaFileManager.class.isAssignableFrom(method.getDeclaringClass())) {
+            throw new IllegalArgumentException(("not a method on JavaFileManager: " + method));
+        }
+
+        var prev = fileManagerHandlers.put(method, handler);
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java:100]: Done.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+     */
+    public TestJavaFileManagerBuilder handle(Predicate<JavaFileObject> filter,
+                                             Method method,
+                                             BiFunction<JavaFileObject, Object[], Throwable> handler) {
+        if (!FileObject.class.isAssignableFrom(method.getDeclaringClass())) {
+            throw new IllegalArgumentException(("not a method on FileObject: " + method));
+        }
+
+        var map = fileObjectHandlers.computeIfAbsent(filter, p_ -> new HashMap<>());
+        var prev = map.put(method, handler);
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java]: I've modified the API so that it is no longer necessary to use the predicate as a key.

Instead, the map is no longer built incrementally by a series of `handle` calls;  the signature of the `handle` call is modified to take the predicate and a user-provided map.  It is expected that maps will be small (typically one element) and easy to create with `Map.of`.    While this specifically applies to the handling of file objects, for consistency a similar change is made to the API to register handlers for file manager calls.

Also of note, because this removes the need to create the maps within the builder class, the other issue about using `putIfAbsent` instead of `put` is now moot.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/resources/javadoc.properties [hunk 7]
```diff
@@ -302,6 +302,8 @@ main.warnings.Werror=warnings found and -Werror specified
 main.unknown.error=an unknown error has occurred
 main.internal.error=an internal error has occurred
 main.unexpected.exception=an unexpected exception was caught: {0}
+main.unknown.option.for.filemanager=option not supported by file manager: {0}
+main.bad.arg.for.filemanager.option=bad value for file manager option {0}: "{1}"
 doclet.internal.report.bug=\
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/resources/javadoc.properties:307]: We don't seem to have any tests for these new errors.

File: test/langtools/jdk/javadoc/testTFMBuilder/TestTFMBuilder.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import javadoc.tester.JavadocTester;
+import javadoc.tester.TestJavaFileManagerBuilder;
+import toolbox.ToolBox;
+
+/**
+ * Tests the {@link TestJavaFileManagerBuilder class}.
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/testTFMBuilder/TestTFMBuilder.java:53]: This doc comment reads weird when rendered.

**Review by pavelrappo — APPROVED** [review]
I can work with that to test snippets; thanks. Update the copyright years before integrating.

File: test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+ * However, it does mean that methods on file objects called internally by a
+ * file manager will not throw any user-provided exceptions.
+ *
+ * For now, the handlers for a file object are simply grouped by predicate and then by
+ * method, and the group of methods used for a "test file object" is determined by the
+ * first predicate that matches.
+ * An alternative, more expensive, implementation would be to group the handlers
+ * by method and predicate and then dynamically build the set of methods to be used for
+ * a file object by filtering the methods by their applicable predicate.
+ */
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java:65]: > [snip]. Update the copyright years before integrating.

Done

[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java:65]: > These 1.5 paragraphs seem no longer relevant.

I think the comments are still useful/relevant because the most recent change was just about avoid the use of a predicate as a map key, but the ordered list off predicates is still there, albeit in a slightly different form.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/resources/javadoc.properties (hunk 7 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/resources/javadoc.properties:307]: Filed https://bugs.openjdk.java.net/browse/JDK-8279251

File: test/langtools/jdk/javadoc/testTFMBuilder/TestTFMBuilder.java (hunk 8 above)
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/testTFMBuilder/TestTFMBuilder.java:53]: Misplaced `}`; will fix.
